**The purpose of analysis** to help the startup o develop new apps for book lovers.

**The task** to get information from the given database of one of the services competing in this market. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition for a new product.

Let's load libraries and make connection with database.

In [1]:
import pandas as pd
import warnings
warnings.filterwarnings("ignore")
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity='all'

In [2]:
!pip install psycopg2-binary

In [3]:
from sqlalchemy import create_engine


db_config = {'user': 'praktikum_student',         # user name
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

Study the tables (print the first rows).

Description of the `**books**` table.
Contains data on books:
- `book_id`
- `author_id`
- `title`
- `num_pages` — number of pages
- `publication_date`
- `publisher_id`

In [4]:
query="""
SELECT * 
FROM books 
"""
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [5]:
books=pd.read_sql_query(query, con= engine)
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [6]:
books.info()
books.duplicated().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


0

- the book table has 1000 rows, 6 columns, without missing values and duplicates.

Description of the `**authors**` table.
Contains data on authors:
- `author_id`
- `author`

In [7]:
query="""
SELECT * 
FROM authors 
"""
authors=pd.read_sql_query(query, con= engine)
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [8]:
authors.info()
authors.duplicated().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


0

- the authors table has 636 rows, 2 columns, without missing values and duplicates.

Description of the `**publishers**` table.
Contains data on publishers:
- `publisher_id`
- `publisher`

In [9]:
query="""
SELECT * 
FROM publishers 
"""
publishers=pd.read_sql_query(query, con= engine)
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [10]:
publishers.info()
publishers.duplicated().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


0

- the publishers table has 340 rows, 2 columns, without missing values and duplicates.

Description of the `**ratings**` table.
Contains data on ratings:
- `rating_id`
- `book_id`
- `username` — the name of the user who rated the book
- `rating`

In [11]:
query="""
SELECT * 
FROM ratings 
"""
ratings=pd.read_sql_query(query, con= engine)
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [12]:
ratings.info()
ratings.duplicated().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


0

- the ratings table has 6456 rows, 4 columns, without missing values and duplicates.

Description of the `**reviews**` table.
Contains data on reviews:
- `review_id`
- `book_id`
- `username` — the name of the user who reviewed the book 
- `text` — the text of the review

In [13]:
query="""
SELECT * 
FROM reviews 
"""
reviews=pd.read_sql_query(query, con= engine)
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [14]:
reviews.info()
reviews.duplicated().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


0

- the reviews table has 6456 rows, 4 columns, without missing values and duplicates.

### Task
#### - Find the number of books released after January 1, 2000.

In [15]:
query="""
SELECT COUNT (DISTINCT book_id) AS number_of_books
FROM books
WHERE CAST(publication_date AS timestamp) BETWEEN '2000-01-01' AND CURRENT_DATE
"""
table1=pd.read_sql_query(query, con= engine)
table1

number_of_books
0              821

- *the number of books released after January 1, 2000 is 821.*

#### - Find the number of user reviews and the average rating for each book.

In [16]:
query="""
SELECT b.title,
    re.number_of_user_reviews,
    r.average_rating
FROM (SELECT book_id,
            title
    FROM books) AS b 
LEFT JOIN 
    (SELECT COUNT(text) AS number_of_user_reviews,
            book_id
    FROM reviews
    GROUP BY book_id) AS re ON b.book_id=re.book_id
LEFT JOIN    
    (SELECT AVG(rating) AS average_rating,
            book_id
        FROM ratings
        GROUP BY book_id) AS r ON b.book_id=r.book_id
ORDER BY r.average_rating DESC
"""
table2=pd.read_sql_query(query, con= engine)
table2

title  \
0                 Pop Goes the Weasel (Alex Cross  #5)   
1    The Ghost Map: The Story of London's Most Terr...   
2    In the Hand of the Goddess (Song of the Liones...   
3                             Tai-Pan (Asian Saga  #2)   
4    How to Be a Domestic Goddess: Baking and the A...   
..                                                 ...   
995  The World Is Flat: A Brief History of the Twen...   
996                                              Junky   
997                  His Excellency: George Washington   
998                                      Drowning Ruth   
999                               Harvesting the Heart   

     number_of_user_reviews  average_rating  
0                       2.0            5.00  
1                       2.0            5.00  
2                       2.0            5.00  
3                       2.0            5.00  
4                       1.0            5.00  
..                      ...             ...  
995                     3.0            2.25  
996                     2.0            2.00  
997                     2.0            2.00  
998                     3.0            2.00  
999                     2.0            1.50  

[1000 rows x 3 columns]

#### - Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).

In [17]:
query="""
SELECT DISTINCT p.publisher,
    COUNT(b.book_id) AS count
FROM publishers AS p
INNER JOIN books AS b ON p.publisher_id=b.publisher_id
WHERE b.num_pages>50
GROUP BY p.publisher
ORDER BY count DESC
LIMIT 1
"""
table3=pd.read_sql_query(query, con= engine)
table3

publisher  count
0  Penguin Books     42

- *the publisher that has released the greatest number of books with more than 50 pages is Penguin Books. They released 42 books.*

#### - Identify the author with the highest average book rating (look only at books with at least 50 ratings).

In [18]:
query="""
SELECT a.author,
    AVG(r.rating)
FROM authors AS a
INNER JOIN books AS b ON a.author_id=b.author_id
INNER JOIN ratings AS r ON b.book_id=r.book_id
GROUP BY a.author
HAVING COUNT(r.rating)>50
ORDER BY AVG(r.rating) DESC
LIMIT 1
"""
table4=pd.read_sql_query(query, con= engine)
table4

author       avg
0  J.K. Rowling/Mary GrandPré  4.288462

- *the author with the highest average book rating is J.K. Rowling/Mary GrandPré (average rating is 4.3).*

#### - Find the average number of text reviews among users who rated more than 50 books.

In [19]:
query="""
SELECT AVG(r.count)
FROM (SELECT COUNT(text) as count
        FROM reviews
        WHERE username IN (SELECT username
                        FROM ratings
                        GROUP BY username
                        HAVING COUNT(rating)>50) 
        GROUP BY username) AS r
"""
table5=pd.read_sql_query(query, con= engine)
table5

avg
0  24.333333

- *The average number of text reviews among users who rated more than 50 books is 24.3.*

**Conclsuion:**

Based on information from the given database of one of the services competing in this market we can see:
- the number of books released after January 1, 2000 is 821.
- the publisher that has released the greatest number of books with more than 50 pages is Penguin Books. They released 42 books.
- the author with the highest average book rating is J.K. Rowling/Mary GrandPré (average rating is 4.3).
- the average number of text reviews among users who rated more than 50 books is 24.3.